tensorflow-ops/src/TensorFlow/Logging.hs

```diff
@@ -0,0 +1,158 @@
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+-- | TensorBoard Summary generation. Provides type safe wrappers around raw
+-- string emitting CoreOps.
+--
+-- Example use:
+--
+-- > -- Call summary functions while constructing the graph.
+-- > createModel = do
+-- >   loss <- -- ...
+-- >   TF.scalarSummary loss
+-- >
+-- > -- Write summaries to an EventWriter.
+-- > train = TF.withEventWriter "/path/to/logs" $ \eventWriter -> do
+-- >     summaryTensor <- TF.build TF.allSummaries
+-- >     forM_ [1..] $ \step -> do
+-- >         if (step % 100 == 0)
+-- >             then do
+-- >                 ((), summaryBytes) <- TF.run (trainStep, summaryTensor)
+-- >                 let summary = decodeMessageOrDie (TF.unScalar summaryBytes)
+-- >                 TF.logSummary eventWriter step summary
+-- >             else TF.run_ trainStep
+
+{-# LANGUAGE TypeOperators #-}
+
+module TensorFlow.Logging
+    ( EventWriter
+    , withEventWriter
+    , logEvent
+    , logSummary
+    , SummaryTensor
+    , histogramSummary
+    , scalarSummary
+    , mergeAllSummaries
+    ) where
+
+import Control.Concurrent (forkFinally)
+import Control.Concurrent.MVar (MVar, newEmptyMVar, readMVar, putMVar)
+import Control.Concurrent.STM (atomically)
+import Control.Concurrent.STM.TBMQueue (TBMQueue, newTBMQueueIO, closeTBMQueue, writeTBMQueue)
+import Control.Monad.Catch (MonadMask, bracket)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Control.Monad.Trans.Resource (runResourceT)
+import Data.ByteString (ByteString)
+import Data.Conduit ((=$=))
+import Data.Conduit.TQueue (sourceTBMQueue)
+import Data.Default (def)
+import Data.Int (Int64)
+import Data.ProtoLens (encodeMessage)
+import Data.Time.Clock (getCurrentTime)
+import Data.Time.Clock.POSIX (utcTimeToPOSIXSeconds)
+import Lens.Family2 ((.~), (&))
+import Network.HostName (getHostName)
+import Proto.Tensorflow.Core.Framework.Summary (Summary)
+import Proto.Tensorflow.Core.Util.Event (Event, fileVersion, step, summary, wallTime)
+import System.Directory (createDirectoryIfMissing)
+import System.FilePath ((</>))
+import TensorFlow.Build (Build, render, SummaryTensor, addSummary, collectAllSummaries)
+import TensorFlow.Ops (scalar)
+import TensorFlow.Records.Conduit (sinkTFRecords)
+import TensorFlow.Tensor (Tensor)
+import TensorFlow.Types (TensorType, type(/=))
+import Text.Printf (printf)
+import qualified Data.ByteString.Lazy as L
+import qualified Data.Conduit as Conduit
+import qualified Data.Conduit.List as Conduit
+import qualified Data.Text as T
+import qualified TensorFlow.GenOps.Core as CoreOps
+
+-- | Handle for logging TensorBoard events safely from multiple threads.
+data EventWriter = EventWriter (TBMQueue Event) (MVar ())
+
+-- | Writes Event protocol buffers to event files.
+withEventWriter ::
+    (MonadIO m, MonadMask m)
+    => FilePath
+    -- ^ logdir. Directory where event file will be written.
```

Comment:
Let's specify explicitly the limitation (which was a reason for me to use FFI internally):
-- ^ logdir. Local filesystem directory where event file will be written.